pkg/foghorn/controller.go

```diff
@@ -1,266 +1,129 @@
-	obj, shutdown := c.queue.Get()
-
-	if shutdown {
-		return false
+	if pluginConfig == nil {
+		pluginConfig = &plugins.ConfigAgent{}
 	}
 
-	// We wrap this block in a func so we can defer c.workqueue.Done.
-	err := func(obj interface{}) error {
-		// We call Done here so the workqueue knows we have finished
-		// processing this item. We also must remember to call Forget if we
-		// do not want this work item being re-queued. For example, we do
-		// not call Forget if a transient error occurs, instead the item is
-		// put back on the workqueue and attempted again after a back-off
-		// period.
-		defer c.queue.Done(obj)
-		var key string
-		var ok bool
-		// We expect strings to come off the workqueue. These are of the
-		// form namespace/name. We do this as the delayed nature of the
-		// workqueue means the items in the informer cache may actually be
-		// more up to date that when the item was initially put onto the
-		// workqueue.
-		if key, ok = obj.(string); !ok {
-			// As the item in the workqueue is actually invalid, we call
-			// Forget here else we'd go into a loop of attempting to
-			// process a work item that is invalid.
-			c.queue.Forget(obj)
-			c.logger.Warnf("expected string in workqueue but got %#v", obj)
-			return nil
-		}
-		// Run the syncHandler, passing it the namespace/name string of the
-		// PipelineActivity resource to be synced.
-		if err := c.syncHandler(key); err != nil {
-			// Put the item back on the workqueue to handle any transient errors.
-			c.queue.AddRateLimited(key)
-			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+	if configMapWatcher == nil {
+		var err error
+		configMapWatcher, err = watcher.SetupConfigMapWatchers(ns, jobConfig, pluginConfig)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create ConfigMap watcher")
 		}
-		// Finally, if no error occurs we Forget this item so it does not
-		// get queued again until another change happens.
-		c.queue.Forget(obj)
-		c.logger.Infof("Successfully synced '%s'", key)
-		return nil
-	}(obj)
-
-	if err != nil {
-		c.logger.WithError(err).Error("failure reconciling")
-		return true
 	}
 
-	return true
+	return &LighthouseJobReconciler{
+		client:           client,
+		scheme:           scheme,
+		logger:           logger,
+		ns:               ns,
+		jobConfig:        jobConfig,
+		pluginConfig:     pluginConfig,
+		ConfigMapWatcher: configMapWatcher,
+		wg:               &sync.WaitGroup{},
+	}, nil
 }
 
-// syncHandler compares the actual state with the desired, and attempts to
-// converge the two.
-func (c *Controller) syncHandler(key string) error {
-	// Convert the namespace/name string into a distinct namespace and name
-	namespace, name, err := cache.SplitMetaNamespaceKey(key)
-	if err != nil {
-		c.logger.Warnf("invalid resource key: %s", key)
-		return nil
-	}
+// SetupWithManager sets up the reconcilier with it's manager
+func (r *LighthouseJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
```

Comment:
I am a bit confused here, why is this a dedicated function? I guess what I am wondering about is why `NewLighthouseJobReconciler` does not just receive a `ctrl.Manager`? It contains the client and scheme and can then be directly linked. Is this just for testing purposes? Or maybe I don't understand the controller-runtime API yet?

Reply:
Probably mainly for testing and separation of concerns.

The reconciler represents the reconciliation loop, it could be a single function (controller-runtime supports that https://github.com/kubernetes-sigs/controller-runtime/blob/3715e20fb99cc4e71ed6b204daf00e7818fe396a/pkg/reconcile/reconcile.go#L96-L102).

`SetupWithManager` is responsible of creating the controller out of a reconciler.

I agree that there's probably some room for improvement here, either in terms of naming and responsibility.